Инкремент 13

Fairness2 opened the pull request:
Добавил работу с ошибками. Улучшена устойчивость отправки метрик путем добавления повторной отправки при ошибках соединения с сервером или с базой данных.

Fairness2 committed:
iter10 | Мок стор для skill

Fairness2 committed:
Merge branch 'refs/heads/iter9' into iter10

# Conflicts:
#	cmd/skill/main.go
#	cmd/skill/main_test.go

Fairness2 committed:
iter10 | Добавил поддержку подключения к базе данных и обработчик ping

Добавлены функции инициализации и закрытия базы данных, обновлены настройки конфигурации для включения DSN базы данных, а также добавлен Dockerfile и docker-compose.yml для контейнеризации. Добавлен обработчик проверки состояния соединения с базой данных через endpoint /ping.

Измененные файлы:
- cmd/server/handlers/ping/handler.go
- deployments/docker/go/Dockerfile
- cmd/server/main.go
- cmd/skill/app.go
- docker-compose.yml
- go.mod
- cmd/server/config/config.go
- go.sum
- internal/metrics/mock/storage.go
- cmd/server/config/parse.go
- cmd/skill/main.go
- internal/database/pool.go

Fairness2 committed:
iter10 | Обновили значение конфигурации по умолчанию для DSN и добавили проверку DSN в пуле базы данных`

Обновили значение конфигурации по умолчанию для переменной `DefaultDatabaseDSN`, закомментировав строку подключения. Добавили проверку пустого значения DSN и возвращаем неработоспособный пул в случае отсутствия DSN в `pool.go`.

Список измененных файлов:
1. cmd/server/config/config.go
2. internal/database/pool.go

Fairness2 committed:
iter10 | Поправил ожидаемое тело в тесте скила

Измененные файлы:
- cmd/skill/main_test.go

Fairness2 committed:
iter11 | Добавили поддержку хранения метрик в базе данных

Добавили реализацию хранилища метрик в базе данных с миграциями для создания таблиц. Методы SetGauge и AddCounter теперь возвращают ошибки, добавлена обработка ошибок в местах вызова этих методов.

Измененные файлы:
- cmd/server/handlers/getmetric/urlhandler_test.go
- cmd/server/main.go
- internal/metrics/MemStorage.go
- go.mod
- cmd/server/handlers/handlemetric/service.go
- internal/metrics/Metrics.go
- go.sum
- internal/metrics/MemStorage_test.go
- internal/metrics/FileStorage.go
- cmd/server/migrations.go
- cmd/server/handlers/getmetric/jsonhandler_test.go
- cmd/server/handlers/getmetrics/handler.go
- internal/metrics/DBStorage.go
- internal/metrics/storage.go

Fairness2 committed:
iter11 | Добавил проверку ошибок rows.Err()

Улучшена обработка ошибок при итерации по строкам в методах получения данных. Добавлены проверки ошибок после выполнения запросов для таблиц `t_gauge` и `t_counter`.

Changed files:
- internal/metrics/DBStorage.go

Fairness2 committed:
iter11 | Добавлен defer для закрытия строк в DBStorage

Изменённые файлы:
- `internal/metrics/DBStorage.go`

Fairness2 committed:
iter12 | Добавлены методы массового обновления метрик и новые эндпоинты

В этом коммите добавлены новые методы для массового обновления метрик. Также добавлены новые эндпоинты для обработки множества метрик через один POST-запрос. В коде появились новые тесты для проверки корректности работы этих изменений.

Список изменённых файлов:
- cmd/server/handlers/handlemetric/jsonmanyhandler_test.go
- cmd/server/main.go
- cmd/skill/flags.go
- cmd/skill/app.go
- internal/metrics/MemStorage.go
- cmd/server/handlers/handlemetric/service.go
- cmd/skill/main.go
- internal/metrics/FileStorage.go
- cmd/server/migrations.go
- internal/store/skill/pg/store.go
- cmd/server/handlers/handlemetric/jsonmanyhandler.go
- internal/models/metrics.go
- internal/metrics/DBStorage.go
- cmd/agent/collector/sender/sender.go

Fairness2 committed:
iter12.1 | Удалены неиспользуемые файлы и улучшена обработка транзакций

В этом коммите удалены неиспользуемые файлы. Улучшена обработка транзакций в нескольких местах, добавлены проверки на ошибки и корректная обработка транзакций в функциях Rollback и Commit.

Список изменённых файлов:
- internal/models/metrics.go
- cmd/skill/app.go
- internal/store/skill/pg/store.go
- internal/metrics/DBStorage.go
- cmd/agent/collector/sender/sender_test.go

Fairness2 committed:
iter12.2 | Исправлена обработка ошибок при откате транзакций`

Обновлены сообщения об ошибках для откатов транзакций в базе данных с целью обеспечения корректного сравнения. Замена строки "transaction has already been committed or rolled back" на "sql: transaction has already been committed or rolled back".

Изменённые файлы:
- internal/metrics/DBStorage.go

Fairness2 committed:
sql: transaction has already been committed or rolled back | Исправлена обработка ошибок при откате транзакций`

Обновлены сообщения об ошибках для откатов транзакций в базе данных с целью обеспечения корректного сравнения. Замена строки "transaction has already been committed or rolled back" на "sql: transaction has already been committed or rolled back".

Изменённые файлы:
- internal/metrics/DBStorage.go

Fairness2 committed:
iter12.3 | Добавлена логика обновления счетчика при добавлении дельты

Обновлена логика обработки счётчиков для корректного добавления значения дельты. Теперь при повторной записи учитывается предыдущее значение счетчика.

Изменённые файлы:
- cmd/server/handlers/handlemetric/service.go

Fairness2 committed:
Merge remote-tracking branch 'origin/iter12' into iter12

Fairness2 committed:
iter13 | Улучшена устойчивость отправки метрик

Улучшена устойчивость отправки метрик путем добавления повторной отправки при ошибках соединения с сервером или с базой данных.

Изменённые файлы:
- cmd/skill/app.go
- internal/store/skill/pg/store.go
- go.mod
- cmd/server/config/config.go
- internal/logger/middleware.go
- go.sum
- internal/store/skill/store.go
- internal/metrics/DBStorage.go
- internal/store/skill/mock/store.go
- cmd/agent/collector/sender/sender.go
- internal/metricerrors/errors.go

Fairness2 committed:
iter13.1 | Исправлены типы интервалов и логгирование запросов

Обновлены типы интервалов с time.Duration на int64 для упрощения работы с конфигурацией и исправлено логирование HTTP запросов.

Изменённые файлы:
- cmd/server/main.go
- cmd/agent/collector/collection/initializer.go
- cmd/server/config/config.go
- internal/logger/middleware.go
- cmd/agent/config/env.go
- cmd/agent/collector/sender/sender.go
- cmd/server/config/parse.go
- internal/metrics/FileStorage.go
- cmd/agent/config/parse.go

Fairness2 committed:
iter13.2 | Вернул метод и путь к запросу в логирование HTTP-запросов

Измененные файлы:
- internal/logger/middleware.go

Fairness2 committed:
iter13.3 | Заменил sync.WaitGroup на errgroup.Group и добавил обработку ошибок в рантайме

Перешел с `sync.WaitGroup` на `errgroup.Group` для улучшения обработки ошибок при завершении горутин. Упростил код, избавившись от лишних горутин в процессе синхронизации метрик и закрытия сервера.

Список измененных файлов:
- cmd/server/main.go
- internal/metrics/FileStorage.go

Fairness2 committed:
iter13.4 | Обработал ошибку закрытия файла в FileStorage.

Добавил обработку ошибки при закрытии файла в методе сохранения метрик. Это предотвратит потерю информации об ошибках, возникающих при закрытии файла.

Список измененных файлов:
internal/metrics/FileStorage.go

vokinneberg commented on cmd/agent/collector/sender/sender.go:
```diff
@@ -60,74 +61,73 @@ func (c *Client) PeriodicSender(ctx context.Context) {
+
 // sendMetrics Функция прохода по метрикам и запуск их отправки
-func (c *Client) sendMetrics() {
+func (c *Client) sendMetrics() error {
```
👍 метод стал гораздо проще!

vokinneberg commented on cmd/agent/collector/sender/sender.go:
```diff
@@ -60,74 +61,73 @@ func (c *Client) PeriodicSender(ctx context.Context) {
-	log.Printf("Sending metric %s with value %d, delta %d type %s\n", body.ID, body.Value, body.Delta, body.MType)
+// sendToServer Отправка метрики
+func (c *Client) sendToServer(body []payload.Metrics) error {
+	log.Println("Sending metrics")
```
Было бы здорово попробовать все-таки использовать структурное логирование - `slog` или `zap`.

Fairness2 replied:
На агенте сделал такой же лог, как на сервере

vokinneberg commented on cmd/server/config/config.go:
```diff
@@ -42,7 +45,8 @@ func InitializeDefaultConfig() *CliConfig {
 		FileStorage:   DefaultFilePath,
 		Restore:       DefaultRestore,
-		StoreInterval: DefaultStoreInterval,
+		StoreInterval: int64(DefaultStoreInterval),
```
Не очень красиво, как пом мне делать приведение здесь. Может быть стоит сразу определить тип?

Fairness2 replied:
Определил дефолтную переменную как int64

vokinneberg commented on cmd/server/handlers/handlemetric/jsonmanyhandler_test.go:
```diff
@@ -0,0 +1,93 @@
+
+			res, err := request.Send()
+			assert.NoError(t, err, "error making HTTP request")
+			assert.Equal(t, test.wantStatus, res.StatusCode())
```
Стоит добавить информативное сообщение к ассертам с информацией о том, что ожидалось и что пошло не так.

Fairness2 replied:
Добавил описание что произошло

vokinneberg commented on cmd/server/handlers/handlemetric/service.go:
```diff
@@ -39,15 +40,23 @@ func updateMetricByStringValue(metricType, metricName, metricValue string) (stri
+		err = metrics.MeStore.SetGauge(metricName, metrics.Gauge(convertedValue))
+		if err != nil {
+			//log.Println(err)
+			return "", &UpdateMetricError{err, http.StatusInternalServerError}
```
👍 за реализацию кастомных ошибок!

vokinneberg commented on cmd/server/handlers/handlemetric/service.go:
```diff
@@ -76,17 +85,71 @@ func updateMetricByRequestBody(body payload.Metrics) (string, *UpdateMetricError
 	return responseMessage, nil
 }
+
+func updateMetricsByRequestBody(bodies []payload.Metrics) (string, *UpdateMetricError) {
```
Не стоит возвращать конкретный тип ошибки. Стоит реализовать кастомную ошибку так, чтобы она реализовывала интерфейс error и возвращать его, а на вызывающей стороне можно будет использовать errors.As(...) для проверки.
```suggestion
func updateMetricsByRequestBody(bodies []payload.Metrics) (string, error) {
```

Fairness2 replied:
Сделал возвращение ошибки, а в вызывающей функции проверяю через errors.As

vokinneberg commented on cmd/server/handlers/handlemetric/service.go:
```diff
@@ -76,17 +85,71 @@ func updateMetricByRequestBody(body payload.Metrics) (string, *UpdateMetricError
+		return "", &UpdateMetricError{err, http.StatusInternalServerError}
+	}
+
+	return "Metrics successfully updated.", nil
```
 Мне кажется лишним возвращать информационное сообщение, как параметр из функции. Думаю там, где функция вызывается можно просто использовать сообщение для записи в лог, если нужно. Возвращаемые параметры лучше определять функциональными - относящимися к тому, что делает функция. В крайнем случае, если хочется добавить индикацию того, что функция выполнилась успешно, можно добавить  bool параметр.

Fairness2 replied:
Убрал возвращение текста, теперь он ставится вызывающей функцией при нулевой ошибке

vokinneberg commented on cmd/server/handlers/ping/handler.go:
```diff
@@ -0,0 +1,24 @@
+
+// Handler возвращает состояние коннекта к базе данных
+func Handler(response http.ResponseWriter, request *http.Request) {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
```
👍 за установку таймаута.

vokinneberg commented on cmd/server/handlers/ping/handler.go:
```diff
@@ -0,0 +1,24 @@
+		helpers.SetHTTPError(response, http.StatusInternalServerError, []byte{})
+		return
+	}
+	helpers.SetHTTPError(response, http.StatusOK, []byte{})
```
У меня название этого метода не бьется с тем, что он делает. Наверное стоит дать ему более информативное название?
```suggestion
	helpers.SetHTTPReponse(response, http.StatusOK, []byte{})
```

Fairness2 replied:
Функция переросла первоначальный смысл, переименовал

vokinneberg commented on cmd/server/main.go:
```diff
@@ -29,43 +31,57 @@ func main() {
 
 	// стартуем приложение
-	if err = runApplication(&wg); err != nil {
+	if err = runApplication(wg); err != nil {
```
На мой взгляд не совсем удачное использование `errgroup.Group`, так как не видно, как она вызывается. Мне кажется стоит сделать вызовы видимым, либо использовать `wg.Wait()` там, где вызывается `Go(...)` и адаптировать код для этого.
```suggestion
	wg.Go(func() error { return runApplication() })
```

Fairness2 replied:
Сделал использование группы только в одном методе

vokinneberg commented on cmd/server/main.go:
```diff
@@ -156,11 +182,9 @@ func InitStore(ctx context.Context, wg *sync.WaitGroup) {
-				store.Sync(ctx)
-			}()
+			wg.Go(func() error {
+				return store.Sync(ctx)
```
Мне кажется по условию задачи DB сторадж должен работать так же как и файловый в смысле синхронизации. Кажется для  DB реализовано только синхронное сохранение?

Fairness2 replied:
Конечно про это ничего не сказано в задании. Но для усложнения задачи сделал)

vokinneberg commented on cmd/skill/app.go:
```diff
@@ -0,0 +1,243 @@
+	return &app{store: s}
+}
+
+func (a *app) webhook(response http.ResponseWriter, request *http.Request) {
```
Метод получился безумно длинный. Стоит подумать о том чтобы разбить его на несколько более коротких методов. Кажется быть стоит рассмотреть возможность использования паттерна visitor или command?

Fairness2 replied:
Разбил обработки на подфункции

vokinneberg commented on cmd/skill/app.go:
```diff
@@ -0,0 +1,243 @@
+	}
+	logger.Log.Debug("sending HTTP 200 response")
+
+	/*text := "Для вас нет новых сообщений."
```
Очень длинный метод, и закомментированный код делает чтение еще сложнее.

Fairness2 replied:
Удалил закомментированный код

vokinneberg commented on internal/metrics/DBStorage.go:
```diff
@@ -0,0 +1,362 @@
+}
+
+// NewDBStorage создание нового хранилища в базе данных
+func NewDBStorage(ctx context.Context, db *sql.DB) (*DBStorage, error) {
```
Думаю, раз мы определили интерфейс мы хотим его принимать и как параметр тоже?
```suggestion
func NewDBStorage(ctx context.Context, db SQLExecutor) (*DBStorage, error) {
```

Fairness2 replied:
Сделал принимающий параметр интерфейсом

vokinneberg commented on internal/metrics/DBStorage.go:
```diff
@@ -0,0 +1,362 @@
+	return g, ok
+}
+
+// Get Получение значения метрики из хранилища
```
Думаю стоит либо определить godoc комментарии для всех методов либо не определять их вовсе, в противном случае как-то не аккуратно.

Fairness2 replied:
Поправил комментарии

vokinneberg commented on internal/metrics/FileStorage.go:
```diff
@@ -47,10 +47,12 @@ func (storage *DurationFileStorage) Sync(ctx context.Context) {
 			ticker.Stop()
 			if err := storage.Flush(); err != nil {
-				logger.Log.Error(err)
+				logger.Log.Debug("Error while syncing metrics")
```
Есть ли причина почему это `Debug`? Думаю ошибку стоит залогировать хотя бы на уровне `Warn`?

Fairness2 replied:
СОгласен, переделал на Warn

vokinneberg commented on internal/metrics/Metrics.go:
```diff
@@ -18,6 +22,26 @@ func (g Gauge) GetRaw() float64 {
+	return g.GetRaw(), nil
+}
+
+func (g *Gauge) Scan(value interface{}) error {
```
Сейчас есть алиас `any` для `interface{}`, мне кажется он предпочтительней.
```suggestion
func (g *Gauge) Scan(value any) error {
```

Fairness2 replied:
Поправил

vokinneberg commented on internal/store/skill/pg/store.go:
```diff
@@ -0,0 +1,174 @@
+// Bootstrap подготавливает БД к работе, создавая необходимые таблицы и индексы
+func (s Store) Bootstrap(ctx context.Context) error {
+	// запускаем транзакцию
+	tx, err := s.conn.BeginTx(ctx, nil)
```
 👍 за исползование транзакции!

vokinneberg reviewed (changes requested):
Ура, первое ревью третьего спринта! 🎉 В этом спринте наше приложение стало еще сложнее, добавился новый слой БД 💪 Ты реализовал доработки сервиса в соответствии с постановкой задачи. Отличная работа! 🚀 Тем не менее мне хотелось бы добавить несколько комментариев относительно твоего кода и сделать несколько предложений по улучшению. 🤓

**Сильные стороны твоего PR:**

1. Код функционален и корректен! 🚀
2. Ты консистентно встроил функционал в структуру существующего приложения. 👍
3. Ты реализовал тесты новых сценариев и дополнил существующие. 👍

**На что я бы рекомендовал обратить внимание и что можно улучшить:**

1. Мне кажется, по условию задачи DB сторадж должен работать так же, как и файловый в смысле синхронизации. Кажется, для DB реализовано только синхронное сохранение? Стоит реализовать асинхронное.
2. Не стоит возвращать конкретный тип ошибки. Стоит реализовать кастомную ошибку так, чтобы она реализовывала интерфейс error и возвращать его, а на вызывающей стороне можно будет использовать [[errors.As](http://errors.as/)](http://errors.as/)(...) для проверки.

Fairness2 committed:
iter13.5 | Доработки по код ревью

Перенес использование логирования на новый модуль и убрал ненужные возвраты сообщений в функциях работы с метриками. В коде работы с БД введена логика синхронизации и восстановления даных, добавлены новые интерфейсы и ошибки.

Список измененных файлов:
- cmd/agent/collector/collection/initializer.go
- cmd/server/handlers/handlemetric/service.go
- internal/metrics/storage.go
- internal/metrics/Metrics.go
- cmd/server/handlers/ping/handler.go
- cmd/agent/collector/sender/sender.go
- internal/middlewares/middlewares.go
- cmd/server/handlers/getmetric/jsonhandler_test.go
- cmd/server/handlers/getmetric/urlhandler_test.go
- internal/metrics/DBStorage.go

Fairness2 committed:
iter13.6 | Фиксы по анализу кода

Добавлены функции для логирования ошибок при вызове defer Close. Удалены неиспользуемые функции PrintConfig в файлах конфигурации.

Список изменённых файлов:
- internal/middlewares/middlewares.go
- cmd/agent/config/env.go
- cmd/server/config/config.go
- cmd/server/handlers/handlemetric/urlhandler.go
- internal/helpers/helpers_test.go
- cmd/server/main.go

Fairness2 committed:
iter13.7 | Обновлены SQL-запросы и переменные

Изменены SQL-запросы для очистки таблиц и переименованы переменные для улучшения читаемости кода. Применён TRUNCATE вместо DELETE для оптимизации производительности.

Список изменённых файлов:
- internal/metrics/DBStorage.go

Fairness2 committed:
iter13.8 | Переименован StorageDatabaseClosed в ErrorStorageDatabaseClosed

Переименован StorageDatabaseClosed в ErrorStorageDatabaseClosed для более точного описания ошибки закрытия базы данных. Это изменение улучшает читаемость и поддерживаемость кода.

Список изменённых файлов:
- internal/metrics/DBStorage.go

vokinneberg reviewed (approved):
Поздравляю с успешным ревью третьего спринта! 🙌  Отличная работа! 🚀 
